executor/plan_recreator.go

```diff
@@ -0,0 +1,177 @@
+)
+
+const recreatorPath string = "/tmp/recreator"
+
+// TTL of plan recreator files
+const remainedInterval float64 = 3
+
+// PlanRecreatorInfo saves the information of plan recreator operation.
+type PlanRecreatorInfo interface {
+	// Process dose the export/import work for reproducing sql queries.
+	Process() (string, error)
+}
+
+// PlanRecreatorSingleExec represents a plan recreator executor.
+type PlanRecreatorSingleExec struct {
+	baseExecutor
+	info *PlanRecreatorSingleInfo
+}
+
+// PlanRecreatorSingleInfo saves the information of plan recreator operation.
+type PlanRecreatorSingleInfo struct {
+	ExecStmt ast.StmtNode
+	Analyze  bool
+	Load     bool
+	File     string
+	Ctx      sessionctx.Context
+}
+
+type fileInfo struct {
+	StartTime time.Time
+	Token     [16]byte
+}
+
+type fileList struct {
+	FileInfo map[string]fileInfo
+	TokenMap map[[16]byte]string
+}
+
+// planRecreatorVarKeyType is a dummy type to avoid naming collision in context.
+type planRecreatorVarKeyType int
+
+// String defines a Stringer function for debugging and pretty printing.
+func (k planRecreatorVarKeyType) String() string {
+	return "plan_recreator_var"
+}
+
+// planRecreatorFileListType is a dummy type to avoid naming collision in context.
+type planRecreatorFileListType int
+
+// String defines a Stringer function for debugging and pretty printing.
+func (k planRecreatorFileListType) String() string {
+	return "plan_recreator_file_list"
+}
+
+// PlanRecreatorVarKey is a variable key for plan recreator.
+const PlanRecreatorVarKey planRecreatorVarKeyType = 0
+
+// PlanRecreatorFileList is a variable key for plan recreator's file list.
+const PlanRecreatorFileList planRecreatorFileListType = 0
+
+// Next implements the Executor Next interface.
+func (e *PlanRecreatorSingleExec) Next(ctx context.Context, req *chunk.Chunk) error {
+	req.GrowAndReset(e.maxChunkSize)
+	if e.info.ExecStmt == nil {
+		return errors.New("plan Recreator: sql is empty")
+	}
+	val := e.ctx.Value(PlanRecreatorVarKey)
+	if val != nil {
+		e.ctx.SetValue(PlanRecreatorVarKey, nil)
+		return errors.New("plan Recreator: previous plan recreator option isn't closed normally")
+	}
+	e.ctx.SetValue(PlanRecreatorVarKey, e.info)
+	return nil
+}
+
+// Close implements the Executor Close interface.
+func (e *PlanRecreatorSingleExec) Close() error {
+	return nil
+}
+
+// Open implements the Executor Open interface.
+func (e *PlanRecreatorSingleExec) Open(ctx context.Context) error {
+	return nil
+}
+
+// Process dose the export/import work for reproducing sql queries.
+func (e *PlanRecreatorSingleInfo) Process() (string, error) {
+	// TODO: plan recreator load will be developed later
+	if e.Load {
+		return "", nil
+	}
+	return e.dumpSingle()
+}
+
+func (e *PlanRecreatorSingleInfo) dumpSingle() (string, error) {
+	// Create path
+	err := os.MkdirAll(recreatorPath, os.ModePerm)
+	if err != nil {
+		return "", errors.New("plan Recreator: cannot create plan recreator path")
```

Comment:
`plan Recreator` seems a little strange. Shall we use `Plan Recreator` or `plan recreator` or something else?